File: Scaler/Queue/PerfectNumber.js
/*

Problem Description

Given an integer A, you have to find the Ath Perfect Number.

A Perfect Number has the following properties:

It comprises only 1 and 2.
The number of digits in a Perfect number is even.
It is a palindrome number.
For example, 11, 22, 112211 are Perfect numbers, where 123, 121, 782, 1 are not.





Problem Constraints

1 <= A <= 100000



Input Format

The only argument given is an integer A.



Output Format

Return a string that denotes the Ath Perfect Number.



Example Input

Input 1:

 A = 2
Input 2:

 A = 3


Example Output

Output 1:

 22
Output 2:

 1111


Example Explanation

Explanation 1:

First four perfect numbers are:
1. 11
2. 22
3. 1111
4. 1221
Return the 2nd Perfect number.
Explanation 2:

First four perfect numbers are:
1. 11
2. 22
3. 1111
4. 1221
Return the 3rd Perfect number.



 
*/



// using BFS approach and dequeue 
const Dequeue = require('double-ended-queue');

const findAthPerfectNumber_BFS = function(A) {
    const dequeue = new Dequeue();
    dequeue.push("1");
    dequeue.push("2");
    let count = 0;

    for (let i = 0; i < A - 1; i++) {
        const front = dequeue.shift();
        dequeue.push(front + "1");
        dequeue.push(front + "2");
    }
    const result = dequeue.shift();
    const reversed = result.split("").reverse().join("");
    return result + reversed;
}


console.log(findAthPerfectNumber_BFS(2)); // 22
console.log(findAthPerfectNumber_BFS(3)); // 1111
console.log(findAthPerfectNumber_BFS(4));  // 1221


// mix of BFS and queue approach
 const findAthPerfectNumber = function(A) {  
    const dequeue = [];
    dequeue.push("1");
    dequeue.push("2");

    let count = 0;

     while (dequeue.length < A) {
        const front = dequeue[count];
        count += 1;
      
        dequeue.push(front + "1");
        dequeue.push(front + "2");
     }
        const reversed = dequeue[A-1].split("").reverse().join("");
        return  dequeue[A-1] + reversed;
}

// console.log(findAthPerfectNumber(2)); // 22
// console.log(findAthPerfectNumber(3)); // 1111
// console.log(findAthPerfectNumber(4)); // 1221
// console.log(findAthPerfectNumber(5)); // 2112
// console.log(findAthPerfectNumber(6)); // 2222

File: Scaler/Queue/PerfectNumber.java
import java.util.ArrayDeque;
import java.util.Deque; // Efficient implementation of Deque

public class PerfectNumber {

    /**
     * Finds the Ath Perfect Number using Breadth-First Search (BFS)
     * with ArrayDeque.
     *  
     *   @param A The index of the Perfect Number to find (1-based).
     *   @return The Ath Perfect Number as a String.
     */

    public static String findAthPerfectNumber_BFS(int A) {
        if (A <= 0) {
            throw new IllegalArgumentException("A must be a positive integer.");
        }

        // Use Deque interface implemented by ArrayDeque
        // push() in JS maps to offerLast() or addLast() in Deque (add to the end)
        // shift() in JS maps to pollFirst() or removeFirst() in Deque (remove from the front)
        Deque<String> deque = new ArrayDeque<>();
        
        // 1. Initialize the deque with base terms S = "1" and S = "2"
        deque.offerLast("1"); // Equivalent to JS push()
        deque.offerLast("2"); // Equivalent to JS push()

        // 2. Loop A-1 times to discard the first A-1 terms
        for (int i = 0; i < A - 1; i++) {
            // Retrieve and remove the element from the front (Equivalent to JS shift())
            String front = deque.pollFirst(); 
            
            if (front == null) {
                // Should not happen for valid A
                break; 
            }
            
            // Generate and enqueue the next two terms
            deque.offerLast(front + "1");
            deque.offerLast(front + "2");
        }
        
        // 3. The A-th term (S) is now at the front. Retrieve it.
        String result = deque.pollFirst(); // Equivalent to JS shift()

        if (result == null) {
             return "Error: Could not find Ath term.";
        }

        // 4. Construct the palindrome S S^R
        StringBuilder sb = new StringBuilder(result);
        String reversed = sb.reverse().toString();
        
        return result + reversed;
    }

    // Example Usage:
    public static void main(String[] args) {
        // A=2 -> S="2" -> Perfect Number: "22"
        System.out.println("2nd Perfect Number (A=2): " + findAthPerfectNumber_BFS(2)); 
        
        // A=4 -> S="12" -> Perfect Number: "1221"
        System.out.println("4th Perfect Number (A=4): " + findAthPerfectNumber_BFS(4)); 
    }
}